chain-impl-mockchain/src/leadership/genesis.rs

```diff
@@ -0,0 +1,1481 @@
+use crate::block::{BlockDate, Message, SignedBlock};
+use crate::certificate;
+use crate::key::PublicKey;
+use crate::leadership::bft::BftRoundRobinIndex;
+use crate::ledger::Ledger;
+use crate::setting::{self, Settings};
+use crate::transaction::Value;
+use crate::update::ValueDiff;
+
+use chain_addr::Kind;
+use chain_core::property::{self, Block, LeaderSelection, Update};
+
+use rand::{Rng, SeedableRng};
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, RwLock};
+
+#[derive(Debug)]
+pub struct GenesisLeaderSelection {
+    ledger: Arc<RwLock<Ledger>>,
+    settings: Arc<RwLock<Settings>>,
+
+    bft_leaders: Vec<PublicKey>,
+
+    stake_keys: HashMap<PublicKey, StakeKeyInfo>,
+
+    stake_pools: HashMap<PublicKey, StakePoolInfo>,
+
+    /// The stake distribution at the end of the previous epoch.
+    stake_snapshot_n_minus_1: StakeDistribution,
+
+    /// The stake distribution at the end of the previous previous
+    /// epoch.
+    stake_snapshot_n_minus_2: StakeDistribution,
+
+    pos: Pos,
+}
+
+// For each stake pool, the total stake value, and the value for the
+// stake pool members.
+type StakeDistribution = HashMap<PublicKey, (Value, HashMap<PublicKey, Value>)>;
```

Comment:
Instead of a tuple `(Value, HashMap<_, Value>)` can we have a proper type?

Reply:
Done.